lib/syscall_shim/uk_syscall_binary.c

```diff
@@ -111,3 +86,51 @@ void ukplat_syscall_handler(struct uk_syscall_ctx *usc)
 	/* Restore extended register state */
 	ukarch_ectx_load((struct ukarch_ectx *)&usc->ectx);
 }
+
+#if CONFIG_LIBSYSCALL_SHIM_STRACE
+static int uk_binsyscall_strace(struct uk_sysrettab_ctx *sysret_ctx)
+{
+#if CONFIG_LIBSYSCALL_SHIM_STRACE_ANSI_COLOR
+	char prsyscallbuf[512]; /* ANSI color is pretty hungry */
+#else /* !CONFIG_LIBSYSCALL_SHIM_STRACE_ANSI_COLOR */
+	char prsyscallbuf[256];
+#endif /* !CONFIG_LIBSYSCALL_SHIM_STRACE_ANSI_COLOR */
+	int prsyscalllen;
+
+	UK_ASSERT(sysret_ctx);
+
+	/* Do we have any system call context to base our strace on? If not,
+	 * we were probably called by a native system call instead of a binary
+	 * one. We have separate dedicated debug messages for those, therefore
+	 * ignore.
+	 */
+	if (!sysret_ctx->usc)
```

Comment:
Can you make this configurable? We need then only one function for native and bincompat but someone should be able to choose if strace should happen just on one of them or both. You could introduce two boolean options, like `CONFIG_LIBSYSCALL_SHIM_STRACE_NATIVE`
`CONFIG_LIBSYSCALL_SHIM_STRACE_BINARY`

Reply:
The separate debug messages are printed just before a call and are not "pretty printed". In some debugging scenarios, you prefer even to have both messages (one on syscall enter, and another one on syscall exit).

Reply:
That's a really good idea! So basically move `PRINTD` to `sysrettab`?

Reply:
What I meant is that strace on native syscall is something you'd maybe also want to have, so exiting this function earlier under the condition `!sysret_ctx->usc` or `sysret_ctx->usc`, should be configurable. The debug message is only on syscall entrance and not exit and not pretty printed.